### ISSUE TITLE
proxy: Memory leak

### ISSUE THREAD

**Issue by davidcollom** [pomerium/pomerium#1670]
## What happened?
We've noticed an increase in memory utilisation when we update pomerium's config.yaml without restarting it.

## What did you expect to happen?
Memory profile stays relatively 

## How'd it happen?

1. Ran `pomerium` in proxy mode only
2. edit config.yaml
2.1. Add/Remove Policies
3. See memory increase
4. Restart proxy
5. See memory return to a much lower level

This was with around 62 policies defined:
![Screenshot 2020-12-10 at 6 55 43 pm](https://user-images.githubusercontent.com/1504448/101818312-92eb9300-3b1b-11eb-8ca5-a5f0fe3f1aaa.png)
@ ~18:50 we restarted the proxy pod.

We have seen this hit over 6GB of memory before being OOM'd due to pod limits being hit, on restart utilisation is around 250-270MB

## What's your environment like?

- Pomerium version (~v0.11.0-rc1~ v0.11.0):
- GKE: v1.17.12-gke.1504

## What's your config.yaml?

```config.yaml
address: :443
authenticate_service_url: https://XXXXXXXXXXXXXXX
authorize_service_url: https://pomerium-authorize.pomerium.svc.cluster.local
cache_service_url: https://pomerium-cache.pomerium.svc.cluster.local
certificate_authority_file: /pomerium/ca.pem
certificate_file: /pomerium/cert.pem
certificate_key_file: /pomerium/privkey.pem
databroker_storage_type: memory
forward_auth_url: https://XXXXXXXXXXXXXXX
grpc_address: :443
idp_client_id: askljabnskjgbasjfbsajhkfbasjkbvaksjbvjksabnvkjas
idp_provider: oidc
idp_provider_url: https://XXXXXXXXXXXXXXX
idp_scopes:
- openid email groups profile
policy:
- allowed_idp_claims:
    perms:
    - XXXXXXXXXXXXXXX
    - XXXXXXXXXXXXXXX
  from: XXXXXXXXXXXXXXX
  to: XXXXXXXXXXXXXXX
- allowed_idp_claims:
    perms:
    - XXXXXXXXXXXXXXX
    - XXXXXXXXXXXXXXX
  from: XXXXXXXXXXXXXXX
  to: XXXXXXXXXXXXXXX
[..SAME SECTIONS REPEATED 64 TIMES...]
```

**Comment by travisgroth** [comment]
Hi @davidcollom,

Can you go into a little more detail on your change patterns?  Is it a static set of routes with regular small changes, or is it a constant churn of new routes and old ones being deleted, or something else?  

Also, is this usage accumulated over time or is it directly correlated with changes?  How long does it take to grow so substantially?  Can you show that memory usage graph over a longer time frame?

Thanks!

**Comment by travisgroth** [comment]
FWIW, I'm able to reproduce in a few different ways fairly easily.  Investigating further.

**Comment by davidcollom** [comment]
Hey @travisgroth sorry for not getting back to you on this one, last week was a busy week with focus elsewhere.

The usage on this was dynamic based off a teams deployment but was rather 3-4 different policies Bering added/removed every 20-30 minutes.

The main config was being generated by a controller/sidecar and inadvertently writing the config regardless of change (i.e no change, it was re-writing the whole file).

Once we added some hash checking on change and only writing on mismatch, we didn't seem to see this issue as much as initially seen. (Hope that makes sense).


I'm glad that you've been able to reproduce this on your side, if there's anything further that I can help with, please do let me know!

**Comment by travisgroth** [comment]
Thanks @davidcollom.  Appreciate the follow up - this is useful info.  Follow up questions:

- In your case, I assume the memory usage was a slow and steady increase?  
- When those 3-4 policies were being changed, was something making each resulting iteration unique or was it the same identical file content each time?

**Comment by davidcollom** [comment]
1. No, it was a very quick ramp up to the memory limit (but not instantaneous) as we were writing the entire config for each and every policy (at the time I believe there was 60+ policies) in a very short period of time.

2. It was the same identical configuration file being written to disk for each policy defined (as above, approximately 192 policies.

Our solution was to ensure md5 checksums on content and disk to ensure we only write the file to disk once per unique configuration.

This has seen memory usage to hover at about 2gb

**Comment by travisgroth** [comment]
@davidcollom the increasing usage when no changes are made should be fixed by #1732 and shipped in `v0.12`.  Closing this.

We're looking at the total stable memory usage as a related but separate effort.

**Comment by davidcollom** [comment]
Thanks @travisgroth for the heads up. We noticed the related PR and Release of `v0.12` , we've seen some improvement in the memory footprint and great to hear that there's further improvements in the works!

We look forward to seeing them! 👍
